packages/api/src/controllers/cdn-data.ts

```diff
@@ -0,0 +1,272 @@
+import { SQLStatement } from "sql-template-strings";
+
+import { authMiddleware } from "../middleware";
+import { validatePost } from "../middleware";
+import { Response, Router } from "express";
+import { db } from "../store";
+import { FindOptions, FindQuery } from "../store/types";
+import logger from "../logger";
+import uuid from "uuid/v4";
+import { User } from "../schema/types";
+
+const app = Router();
+
+interface CdnUsageRowReq {
+  stream_id: string;
+  playback_id: string;
+  unique_users: number;
+  total_filesize: number;
+  total_cs_bytes: number;
+  total_sc_bytes: number;
+  count: number;
+  // filled in /api/cdn-data handler
+  user_id: string;
+  user_email: string;
+}
+
+interface SendData {
+  date: number;
+  region: string;
+  file_name: string;
+  data: Array<CdnUsageRowReq>;
+}
+
+async function addMany(
+  date: number,
+  region,
+  fileName: string,
+  rows: Array<CdnUsageRowReq>
+): Promise<Error> {
+  const name = "cdn_usage_reg";
+  const client = await db.pool.connect();
+  const newId = uuid();
+  console.log(
+    `-----> add many date=${date} region=${region} fileName=${fileName}`
+  );
+  try {
+    await client.query("BEGIN");
+    for (const tdoc of rows) {
+      await client.query(
+        `INSERT INTO ${name} VALUES (to_timestamp($1), $2, $3, $4, $5, $6, $7, $8, $9, $10)
+          ON CONFLICT (date, region, playback_id)
+          DO UPDATE SET 
+              unique_users = ${name}.unique_users + EXCLUDED.unique_users,
+              total_filesize = ${name}.total_filesize + EXCLUDED.total_filesize,
+              total_cs_bytes = ${name}.total_cs_bytes + EXCLUDED.total_cs_bytes,
+              total_sc_bytes = ${name}.total_sc_bytes + EXCLUDED.total_sc_bytes,
+              count = ${name}.count + EXCLUDED.count;
+          `,
+        [
+          date,
+          region,
+          tdoc.playback_id,
+          tdoc.user_id,
+          tdoc.user_email,
+          tdoc.unique_users,
+          tdoc.total_filesize,
+          tdoc.total_cs_bytes,
+          tdoc.total_sc_bytes,
+          tdoc.count,
+        ]
+      );
+    }
+    await client.query(
+      `INSERT INTO cdn_usage_last VALUES ($1, $2)
+          ON CONFLICT ((data ->> 'region'::text) )
+          DO UPDATE SET 
+            data = EXCLUDED.data
+      `,
+      [newId, JSON.stringify({ region, fileName })]
+    );
+    await client.query("COMMIT");
+  } catch (e) {
+    logger.warn(`--> error, rolling back`);
+    logger.info("--");
+    await client.query("ROLLBACK");
+    logger.warn(`--> error: ${e}`);
+    if (e.message.includes("duplicate key value")) {
+      // throw new BadRequestError(e.detail);
+      // return new BadRequestError(e.detail);
+    }
+    // throw e;
+    return e;
```

Comment:
Would it be possible to throw the error instead of returning? Returning seems very uncommon for js/ts